### ISSUE TITLE
einops cheetsheet

### ISSUE THREAD

**Issue by prakashjayy** [prakashjayy/fullstack_dl#16]
einops is very useful, there is no denying in it for the following reasons. 
- In every CNN classification network, we use `x.view(batch_size, -1)` or `torch.flatten(x, 1)` as [shown here](https://github.com/pytorch/vision/blob/810811f81ca24b196d08778c9e7f50bc5fe3ef3d/torchvision/models/resnet.py#L243) (where x is a tensor of shape (b, c, h, w)) to reshape the tensor so that we can do matrix multiplication with last linear layer. The problem is that we don't know what it is the shape of the code when we look at the code. using einops we can rewrite as `rearrange(ims, 'b c h w -> b (c h w )')`
- suppose u want to transpose the axis from `b, h, w, c` to `b, c, w, h`. with PyTorch we can do it using x.transpose(0, 3, 1, 2) and with einops, we can do `rearrange(x, 'b h w c -> b c w h')`. einops is more clear. 
- when doing squeeze and unsqueeze in PyTorch for decreasing and increasing a dimension we exactly don't know the tensor dimension while reading the code and we keep guessing what the new tensor shape would be, using einops if u want to add a batch dimension to (c, w, h) tensor we do `rearrange(x, '(b) c h w -> b h w c ', b=1).shape` 
- suppose we want to do mean across channels for a tensor named `x`. in pytorch we do x.mean(1) as the shape of x tensor is `b, c, h, w`, but we will not know the shape information when reading the PyTorch code. In einops, we do this by using reduce( `reduce(x, ''b c h w -> b h w', 'mean')`). similarly, we can do min, max and median too. 

## PYTORCH <--> EINOPS 
|einops | PyTorch | Notes|
| --- | --- | -- |
| rearrange | transpose, reshape, stack, concatenate, squeeze and expand_dims | --|
| repeat | repeating and tiling| -- |
| reduce | max, min, mean, median pooling | For example, In Pytorch max-pooling `nn.MaxPool2d(kernel_size=2, stride=2, padding=1)(x)` and einops we do using `reduce(x, 'b (h h2) (w w2) c -> b h w c', 'max', h2=2, w2=2)`|

## Other important stuff
- we can backprop through einops operations.  
- functions are available as layers too. we can import them `from einops.layers.torch import Rearrange, Reduce`.

## Common PyTorch functions written in einops 
| name | einops |
| --- | ---- |
| flatten | `rearrange(x, 'b c h w -> b (c h w)')`|
| global average pooling | `reduce(x, 'b c h w -> b c', reduction='mean')`|
| max pooling | `reduce(x, 'b c (h h1) (w w1) -> b c h w', reduction='max', h1=2, w1=2)`|
| unsqueeze | `rearrange(image, 'h w c -> () c h w')`| 
| stacking | `rearrange(list_of_tensors, 'b c h w -> b h w c')` |
| concatenate | `rearrange(list_of_tensors, 'b c h w -> (b h) w c')`|
| shuffling |  `rearrange(x, 'b c h w-> b h w c')` |

## Pytorch code example 
Using einops we can apply the entire graph in `nn.sequential`. This essentially helps when we port graph from one framework to another framework using something like `onnx`.  Following is one example,

```python
# example given for pytorch, but code in other frameworks is almost identical
from torch.nn import Sequential, Conv2d, MaxPool2d, Linear, ReLU
from einops.layers.torch import Reduce

model = Sequential(
    Conv2d(3, 6, kernel_size=5),
    MaxPool2d(kernel_size=2),
    Conv2d(6, 16, kernel_size=5),
    # combined pooling and flattening in a single step
    Reduce('b c (h 2) (w 2) -> b (c h w)', 'max'), 
    Linear(16*5*5, 120), 
    ReLU(),
    Linear(120, 10), 
)
```

## Resources
- http://einops.rocks/pytorch-examples.html
- [tutorial-1](https://github.com/arogozhnikov/einops/blob/master/docs/1-einops-basics.ipynb)
- [tutorial-2](https://github.com/arogozhnikov/einops/blob/master/docs/2-einops-for-deep-learning.ipynb)
